### PR TITLE
feat: [CDS-86929]: Plan Creator for approval Stage v1

### PR DESCRIPTION
**Pull request by sarthakkasat**
## Describe your changes
Tested my changes by running a pipeline with Custom and Approval Stage with Http Step.

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/56825)

### CODE REVIEW

File: pipeline-service/modules/orchestration-steps/src/main/java/io/harness/plancreator/approval/v1/StepParameterUtils.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+@OwnedBy(PIPELINE)
+@CodePulse(module = ProductModule.CDS, unitCoverageRequired = false, components = {HarnessModuleComponent.CDS_PIPELINE})
+public class StepParameterUtils {
+  public StageElementParametersV1.StageElementParametersV1Builder getStageParameters(ApprovalStageNodeV1 stageNode) {
+    StageElementParametersV1.StageElementParametersV1Builder stageBuilder = getCommonStageParameters(stageNode);
+    stageBuilder.failure(ParameterField.isNotNull(stageNode.getFailure()) ? stageNode.getFailure().getValue() : null);
+    return stageBuilder;
+  }
+
+  private StageElementParametersV1.StageElementParametersV1Builder getCommonStageParameters(
```
Comment (shalini-agr) [inline pipeline-service/modules/orchestration-steps/src/main/java/io/harness/plancreator/approval/v1/StepParameterUtils.java:36]: we have this exact same method in 125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/stage/v1/StepParametersUtils.java 
let's put them in ng-common-utilities and use it from there.

File: pipeline-service/modules/orchestration-steps/src/main/java/io/harness/plancreator/approval/v1/ApprovalStagePlanCreator.java [hunk 2]
```diff
@@ -40,12 +54,63 @@ public ApprovalStageNodeV1 getFieldObject(YamlField field) {
...
+
+    return planCreationResponseMap;
   }
   @Override
   public PlanNode createPlanForParentNode(
       PlanCreationContext ctx, ApprovalStageNodeV1 field, List<String> childrenNodeIds) {
-    return null;
+    StageElementParametersV1.StageElementParametersV1Builder stageParameters =
+        StepParameterUtils.getStageParameters(field);
+    stageParameters.type(YAMLFieldNameConstants.CUSTOM_V1);
```
Comment (shalini-agr) [inline pipeline-service/modules/orchestration-steps/src/main/java/io/harness/plancreator/approval/v1/ApprovalStagePlanCreator.java:89]: this should be APPROVAL_V1

Comment (sarthakkasat) [reply pipeline-service/modules/orchestration-steps/src/main/java/io/harness/plancreator/approval/v1/ApprovalStagePlanCreator.java]: Done

**Review by archit-harness — CHANGES_REQUESTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56825)* status: 8 file(s) to review. 1 change requests. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`125-cd-nextgen/BUILD.bazel` line 149 at r2](https://reviewable.io/reviews/harness/harness-core/56825#-Nlh6Nor3f6PEDXkkmxA:-Nlh6Nor3f6PEDXkkmxB:b-x5pn2o) ([raw file](https://github.com/harness/harness-core/blob/7d73d0a8f753c6e1b739142b34fe34825176e15e/125-cd-nextgen/BUILD.bazel#L149)):*
> ```Starlark
>         "//clients/pipeline-service/pms-sdk:module",
>         "//clients/pipeline-service/pms-sdk-core:module",
>         "//pipeline-service/modules/orchestration-steps:module",
> ```

why do you need this?
___
*[`pipeline-service/modules/orchestration-steps/src/main/java/io/harness/plancreator/approval/v1/StepParameterUtils.java` line 36 at r1](https://reviewable.io/reviews/harness/harness-core/56825#gh-1427720631:-Nlh6VhO2PGYlOOicrIU:b32v57) ([raw file](https://github.com/harness/harness-core/blob/13be19a84a87e2a9cb9540bed4295a8914db5336/pipeline-service/modules/orchestration-steps/src/main/java/io/harness/plancreator/approval/v1/StepParameterUtils.java#L36)):*
<details><summary><i>Previously, shalini-agr (Shalini Agrawal) wrote…</i></summary><blockquote>

we have this exact same method in 125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/stage/v1/StepParametersUtils.java 
let's put them in ng-common-utilities and use it from there.
</blockquote></details>

+1

**Review by sarthakkasat — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56825)* status: 10 file(s) to review. 1 change requests. 3. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`125-cd-nextgen/BUILD.bazel` line 149 at r2](https://reviewable.io/reviews/harness/harness-core/56825#-Nlh6Nor3f6PEDXkkmxA:-NlhOBjB5-BNTRA307Wq:bwk0n31) ([raw file](https://github.com/harness/harness-core/blob/7d73d0a8f753c6e1b739142b34fe34825176e15e/125-cd-nextgen/BUILD.bazel#L149)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

why do you need this?
</blockquote></details>

Removed it, moved the method ng-common-utilities
___
*[`pipeline-service/modules/orchestration-steps/src/main/java/io/harness/plancreator/approval/v1/StepParameterUtils.java` line 36 at r1](https://reviewable.io/reviews/harness/harness-core/56825#gh-1427720631:-NlhOYUeDIzeqmIahVbk:bb6un1j) ([raw file](https://github.com/harness/harness-core/blob/13be19a84a87e2a9cb9540bed4295a8914db5336/pipeline-service/modules/orchestration-steps/src/main/java/io/harness/plancreator/approval/v1/StepParameterUtils.java#L36)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

+1
</blockquote></details>

Done

**Review by archit-harness — CHANGES_REQUESTED** [review]
> Reviewed 1 of 6 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56825)* status: 10 file(s) to review. 1 change requests. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`125-cd-nextgen/BUILD.bazel` line 149 at r3](https://reviewable.io/reviews/harness/harness-core/56825#-NlhRVBZBCIZaLhqQZ2A:-NlhRVBZBCIZaLhqQZ2B:b-e22dd1) ([raw file](https://github.com/harness/harness-core/blob/d6980c9eecfb581903a1cb34e03daf3c50749de1/125-cd-nextgen/BUILD.bazel#L149)):*
> ```Starlark
>         "//clients/pipeline-service/pms-sdk:module",
>         "//clients/pipeline-service/pms-sdk-core:module",
>         "//pipeline-service/modules/orchestration-steps:module",
> ```

i still dont see the remove

**Review by sarthakkasat — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56825)* status: 10 file(s) to review. 1 change requests. 3. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`125-cd-nextgen/BUILD.bazel` line 149 at r3](https://reviewable.io/reviews/harness/harness-core/56825#-NlhRVBZBCIZaLhqQZ2A:-NlhVOwv2I0OZ75VRgwg:bz0qfcb) ([raw file](https://github.com/harness/harness-core/blob/d6980c9eecfb581903a1cb34e03daf3c50749de1/125-cd-nextgen/BUILD.bazel#L149)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

i still dont see the remove
</blockquote></details>

Forgot to commit the file earlier please check again

**Comment by brijeshdhakar** [comment]
*[`pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/approval/stage/v1/ApprovalStageStep.java` line 48 at r5](https://reviewable.io/reviews/harness/harness-core/56825#-NliH4FhC28AROpigts6:-NliH4FhC28AROpigts7:b-prxhti) ([raw file](https://github.com/harness/harness-core/blob/5c0a548a2cba7ac9ab62f292d56a76aa15633d40/pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/approval/stage/v1/ApprovalStageStep.java#L48)):*
> ```Java
>   public ChildExecutableResponse obtainChild(
>       Ambiance ambiance, StageElementParametersV1 stepParameters, StepInputPackage inputPackage) {
>     ApprovalStageSpecParameters specParameters = (ApprovalStageSpecParameters) stepParameters.getSpec();
> ```

Are the ApprovalStageSpecParameters same for V0 and V1? Don't we have any change in approval YAML?

**Review by sarthakkasat — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56825)* status: 9 file(s) to review. 1 change requests. 4. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/approval/stage/v1/ApprovalStageStep.java` line 48 at r5](https://reviewable.io/reviews/harness/harness-core/56825#-NliH4FhC28AROpigts6:-NliVgbg4UE0B9DAM17M:b-qt3ggb) ([raw file](https://github.com/harness/harness-core/blob/5c0a548a2cba7ac9ab62f292d56a76aa15633d40/pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/approval/stage/v1/ApprovalStageStep.java#L48)):*
<details><summary><i>Previously, brijeshdhakar (Brijesh Dhakar) wrote…</i></summary><blockquote>

Are the ApprovalStageSpecParameters same for V0 and V1? Don't we have any change in approval YAML?
</blockquote></details>

We are only using the child node Id from the specParameters, which will the same field for v0, v1. Do we have to create a new class in that case?

**Review by brijeshdhakar — APPROVED** [review]
> Reviewed 2 of 2 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56825)* status: 9 file(s) to review. 1 change requests. 3. unresolved discussions. 1 of 2 approvals obtained.

**Comment by sarthakkasat** [comment]
trigger uts

**Review by shalini-agr — DISMISSED** [review]

**Comment by sarthakkasat** [comment]
trigger review

**Review by archit-harness — CHANGES_REQUESTED** [review]
> Reviewed 3 of 5 files at r1, 2 of 6 files at r3, 1 of 1 files at r4, 2 of 2 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56825)* status: <span>:shipit:</span> complete! ✓

___
*[`pipeline-service/modules/orchestration-steps/src/main/java/io/harness/plancreator/approval/v1/ApprovalStagePlanCreator.java` line 69 at r7](https://reviewable.io/reviews/harness/harness-core/56825#-NlvrmxhAmO-2fU2LfNs:-NlvrmxhAmO-2fU2LfNt:b-ijsfkr) ([raw file](https://github.com/harness/harness-core/blob/b6d05f3aaf9a645c4a55ce8dac8f920d2c83e0ea/pipeline-service/modules/orchestration-steps/src/main/java/io/harness/plancreator/approval/v1/ApprovalStagePlanCreator.java#L69)):*
> ```Java
>     if (stepsField == null) {
>       throw new InvalidRequestException("Steps section is required in Approval stage");
>     }
> ```

why we are checking this? wouldnt this be check with schema validation?



_Code quote:_
```Java
    // Add dependency for execution
    YamlField stepsField = specField.getNode().getField(YAMLFieldNameConstants.STEPS);
    if (stepsField == null) {
      throw new InvalidRequestException("Steps section is required in Approval stage");
    }
```

**Review by archit-harness — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56825)* status: <span>:shipit:</span> complete! ✓

**Comment by sarthakkasat** [comment]
trigger uts

**Comment by sarthakkasat** [comment]
trigger uts

**Review by archit-harness — APPROVED** [review]
> Reviewed 1 of 1 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56825)* status: 1 file(s) to review. 1 of 2 approvals obtained.

**Review by brijeshdhakar — APPROVED** [review]
> Reviewed 1 of 1 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56825)* status: <span>:shipit:</span> complete! ✓

**Comment by sarthakkasat** [comment]
trigger review

**Comment by yogesh-chauhan** [comment]
*[`125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/stage/v1/StepParametersUtils.java` line 27 at r9](https://reviewable.io/reviews/harness/harness-core/56825#-NlyG98z8maXMriJFbCs:-NlyG98z8maXMriJFbCt:b-p20yks) ([raw file](https://github.com/harness/harness-core/blob/c9fd4c4ed03f5d8e5759f19378f5b7554130a5b2/125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/stage/v1/StepParametersUtils.java#L27)):*
> ```Java
> @CodePulse(module = ProductModule.CDS, unitCoverageRequired = false, components = {HarnessModuleComponent.CDS_PIPELINE})
> public class StepParametersUtils {
>   public StageElementParametersV1Builder getStageParameters(DeploymentStageNodeV1 stageNode) {
> ```

should we not call this "getStageParametersBuilder" ? 
or 
this returns StageElementParameters. Callers caller the toBuilder() method to override properties ?

**Comment by yogesh-chauhan** [comment]
*[`878-ng-common-utilities/src/main/java/io/harness/plancreator/stages/v1/StageParameterUtilsV1.java` line 24 at r9](https://reviewable.io/reviews/harness/harness-core/56825#-NlyGQu389cs5OOqs7Bw:-NlyGQu389cs5OOqs7Bx:b-6jg0b9) ([raw file](https://github.com/harness/harness-core/blob/c9fd4c4ed03f5d8e5759f19378f5b7554130a5b2/878-ng-common-utilities/src/main/java/io/harness/plancreator/stages/v1/StageParameterUtilsV1.java#L24)):*
> ```Java
> @UtilityClass
> @OwnedBy(HarnessTeam.PIPELINE)
> public class StageParameterUtilsV1 {
> ```

nit: new line for readability

**Comment by yogesh-chauhan** [comment]
*[`pipeline-service/modules/orchestration-steps/src/main/java/io/harness/plancreator/approval/v1/ApprovalStagePlanCreator.java` line 68 at r9](https://reviewable.io/reviews/harness/harness-core/56825#-NlyH2BK-fm0oRwnBMy5:-NlyH2BK-fm0oRwnBMy6:b-nwovbw) ([raw file](https://github.com/harness/harness-core/blob/c9fd4c4ed03f5d8e5759f19378f5b7554130a5b2/pipeline-service/modules/orchestration-steps/src/main/java/io/harness/plancreator/approval/v1/ApprovalStagePlanCreator.java#L68)):*
> ```Java
>     YamlField stepsField = specField.getNode().getField(YAMLFieldNameConstants.STEPS);
>     if (stepsField == null) {
>       throw new InvalidRequestException("Steps section is required in Approval stage");
> ```

Lets set REPORT Target also as USER

**Comment by yogesh-chauhan** [comment]
*[`pipeline-service/modules/orchestration-steps/src/main/java/io/harness/plancreator/approval/v1/ApprovalStagePlanCreator.java` line 60 at r9](https://reviewable.io/reviews/harness/harness-core/56825#-NlyHABnEGHv0WJ1Y6Ik:-NlyHABnEGHv0WJ1Y6Il:b-boxc8m) ([raw file](https://github.com/harness/harness-core/blob/c9fd4c4ed03f5d8e5759f19378f5b7554130a5b2/pipeline-service/modules/orchestration-steps/src/main/java/io/harness/plancreator/approval/v1/ApprovalStagePlanCreator.java#L60)):*
> ```Java
>       PlanCreationContext ctx, ApprovalStageNodeV1 field) {
>     LinkedHashMap<String, PlanCreationResponse> planCreationResponseMap = new LinkedHashMap<>();
>     Map<String, YamlField> dependenciesNodeMap = new HashMap<>();
> ```

Lets make these final

**Review by yogesh-chauhan — COMMENTED** [review]
> Reviewed 3 of 5 files at r1, 2 of 6 files at r3, 1 of 1 files at r4, 1 of 1 files at r7, 1 of 1 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56825)* status: <span>:shipit:</span> complete! ✓

**Review by yogesh-chauhan — APPROVED** [review]

**Comment by sarthakkasat** [comment]
Will address the comments by @yogesh-chauhan in next PR

**Pull request merged**